### PR TITLE
Allow explicit pre-split pipeline microbatches

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #188500

Pipeline schedules historically treat step() inputs as full batches and split positional arguments, keyword arguments, and targets internally. Input pipelines that already produce microbatches need to bypass that split without changing the schedule-internal representation used by first, middle, and last stages.

Add explicit keyword-only arg_mbs, kwarg_mbs, and target_mbs inputs. Supplying any of them selects pre-split mode, so the API does not need a separate mode flag; even an empty list counts as supplied and reaches the normal microbatch-count validation. Pre-split inputs cannot be mixed with positional inputs, ordinary model kwargs, or target. Positional and keyword microbatch entries are validated together after the existing container and length checks.

Testing coverage:

The single-process matrix directly compares whole-batch and pre-split outputs, losses, and gradients across GPipe, 1F1B, Interleaved1F1B, InterleavedZeroBubble, and LoopedBFS. Focused validation covers mixed input modes, invalid container and entry types, and microbatch-count mismatches. Evaluation coverage compares whole-batch and pre-split results, verifies that evaluation creates no gradients, and verifies that the next training step restores backward execution.

The four-GPU matrix runs both modes through traced GPipe and 1F1B schedules and through manual Interleaved1F1B, InterleavedZeroBubble, and LoopedBFS schedules, checking results and gradients against a non-pipelined reference and retaining tensor-leak coverage. The FlexAttention case uses eight transformer stages with stage-local BlockMask kwargs, matching TorchTitan's pipeline contract. Whole-batch mode lets pipeline parallelism split a full packed-document mask, while pre-split mode constructs each mask independently from its positions microbatch; outputs, losses, and all local gradients are compared directly and against the full-model reference.

Test Plan:

```bash
python -m py_compile test/distributed/pipelining/model_registry.py test/distributed/pipelining/test_schedule.py test/distributed/pipelining/test_schedule_multiproc.py
python test/distributed/pipelining/test_microbatch.py -k block_mask
python test/distributed/pipelining/test_schedule.py -k pre_split
python test/distributed/pipelining/test_schedule.py -k eval_then_train
python test/distributed/pipelining/test_schedule.py
python test/distributed/pipelining/test_schedule_multiproc.py -k interleaved_1f1b_pre_split_flex_attention
python test/distributed/pipelining/test_schedule_multiproc.py -k pre_split
lintrunner -a
```

Authored with an AI assistant.

### CODE REVIEW

File: docs/source/distributed.pipelining.md [hunk 1]
```diff
@@ -110,6 +110,22 @@ else:
...
+arg_mbs = [(x0,), (x1,)]
+kwarg_mbs = [{"mask": mask0}, {"mask": mask1}]
+target_mbs = [target0, target1]
+
+schedule.step(
+    arg_mbs,
+    kwargs=kwarg_mbs,
+    target=target_mbs,
+    pre_split_args_kwargs=True,
+)
```
Comment (fegin) [inline docs/source/distributed.pipelining.md:124]: Claude is pretty upset about this signature and prefer a more explicit one, `step(...., *, arg_mbs=, kwarg_mbs=, target_mbs=)`. The argument is that args is not an args but rather a list of args. Same as kwargs. So it is confusing to reuse `args`, `kwargs`. What's your thought about this?

File: docs/source/distributed.pipelining.md [hunk 2]
```diff
@@ -110,6 +110,22 @@ else:
     output = schedule.step()
 ```
 
+If your input pipeline already produces microbatches, pass them directly and
+set `pre_split_args_kwargs=True`:
+
+```python
+arg_mbs = [(x0,), (x1,)]
+kwarg_mbs = [{"mask": mask0}, {"mask": mask1}]
+target_mbs = [target0, target1]
```
Comment (tianyu-l) [inline docs/source/distributed.pipelining.md:118]: Could you also propose a companying draft PR on torchtitan side? You mentioned that some code could be shared with gradient accumulation, which sounds reasonable.

File: docs/source/distributed.pipelining.md (hunk 1 above)
Comment (sanketpurandare) [reply docs/source/distributed.pipelining.md:124]: @fegin good suggestion, I have updated the PR, please review now.

File: docs/source/distributed.pipelining.md (hunk 2 above)
Comment (sanketpurandare) [reply docs/source/distributed.pipelining.md:118]: @tianyu-l this is the torchtitan draft PR: https://github.com/pytorch/torchtitan/pull/3856. Note, it is not the final state but expresses the idea that I have.

File: torch/distributed/pipelining/schedules.py [hunk 3]
```diff
@@ -615,6 +678,77 @@ def _split_inputs(
...
+
+        if target is not None:
+            raise ValueError(
+                "When pre_split_args_kwargs=True, pass pre-split targets through "
+                "target_mbs=... instead of target=..."
+            )
+
+        arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs)
+
+        for mb_index, arg_mb in enumerate(arg_mbs):
```
Comment (acisseJZhong) [inline torch/distributed/pipelining/schedules.py]: nit: should `arg_mbs` and `kwarg_mbs` always be the same length? if so we could check within one for loop?

File: torch/distributed/pipelining/schedules.py [hunk 4]
```diff
@@ -515,38 +515,101 @@ def step(
...
+
+        ``eval`` uses the same input contract as ``step``, but temporarily
+        disables backward execution. By default, ``args``, ``kwargs``, and
+        ``target`` are full-batch values that the schedule splits into
+        microbatches. Set ``pre_split_args_kwargs=True`` when the caller has
+        already split inputs into microbatches, and pass them through
+        ``arg_mbs``, ``kwarg_mbs``, and ``target_mbs`` instead.
+
+        Args:
+            \*args (Any): Whole-batch positional root inputs when this rank owns
```
Comment (acisseJZhong) [inline torch/distributed/pipelining/schedules.py:546]: Is it possible to only use `arg_mbs`, `kwarg_mbs`, and `target_mbs`? is the main reason not doing this because of BC? 
```
  - pre_split_args_kwargs=False: arg_mbs=[(x,)], kwarg_mbs=[{"mask": m}]
  (length-1) → schedule splits that one element into n_microbatches.
  - pre_split_args_kwargs=True: arg_mbs=[(x0,), (x1,), ...] (length-N) →
  use as-is.
```

Comment (sanketpurandare) [reply torch/distributed/pipelining/schedules.py:546]: yep want to preserve bc for now

File: torch/distributed/pipelining/schedules.py [hunk 5]
```diff
@@ -615,6 +678,77 @@ def _split_inputs(
             # Return a list of empty tuples/dicts with matching length as chunks
             return [()] * self._n_microbatches, [{}] * self._n_microbatches
 
+    def _get_microbatch_inputs(
+        self,
+        args: tuple[Any, ...],
```
Comment (Skylion007) [inline torch/distributed/pipelining/schedules.py:668]: TypeVarTuple here at least?

File: torch/distributed/pipelining/schedules.py [hunk 6]
```diff
@@ -1963,19 +2159,81 @@ def step(
         losses: list | None = None,
         return_outputs: bool = True,
         loss_kwargs: dict[str, Any] | None = None,
+        arg_mbs: Any = None,
+        kwarg_mbs: Any = None,
+        target_mbs: Any = None,
+        pre_split_args_kwargs: bool = False,
```
Comment (fegin) [inline torch/distributed/pipelining/schedules.py]: Do we need this extra argument? Can we infer from `arg_mbs`, `kwarg_mbs`, `target_mbs`?

File: torch/distributed/pipelining/schedules.py [hunk 7]
```diff
@@ -515,38 +515,101 @@ def step(
...
+        self,
+        *args,
+        target=None,
+        losses: list | None = None,
+        arg_mbs: Any = None,
+        kwarg_mbs: Any = None,
+        target_mbs: Any = None,
+        pre_split_args_kwargs: bool = False,
+        **kwargs,
+    ):
```
Comment (fegin) [inline torch/distributed/pipelining/schedules.py:535]: It seems that we don't have an unittest for eval. Let's add one.

**Review by fegin — APPROVED** [review]
Thanks for the refactor. Left two comments.

File: torch/distributed/pipelining/schedules.py (hunk 5 above)
Comment (sanketpurandare) [reply torch/distributed/pipelining/schedules.py:668]: Kept `tuple[Any, ...]` intentionally. Model positional arguments can be heterogeneous, and this helper does not propagate their individual types into a typed return, so a `TypeVarTuple` would not express a useful type relationship here.

File: torch/distributed/pipelining/schedules.py (hunk 7 above)
Comment (sanketpurandare) [reply torch/distributed/pipelining/schedules.py:535]: Extended `test_schedule_eval_then_train` across the five covered schedule implementations. It now compares whole-batch and pre-split evaluation outputs/losses, verifies evaluation produces no parameter gradients, and verifies the next pre-split training step restores backward behavior. The focused eval suite passed all 5 cases.

File: torch/distributed/pipelining/schedules.py (hunk 3 above)
Comment (sanketpurandare) [reply torch/distributed/pipelining/schedules.py]: Combined positional and keyword microbatch entry validation into one `zip(..., strict=True)` loop after the existing container-length checks.

File: torch/distributed/pipelining/schedules.py (hunk 6 above)
Comment (sanketpurandare) [reply torch/distributed/pipelining/schedules.py]: Removed the extra boolean. Supplying any non-`None` `arg_mbs`, `kwarg_mbs`, or `target_mbs` now selects pre-split mode; empty lists also count as supplied and fail the normal microbatch-count validation. With none supplied, the existing whole-batch splitting path is unchanged, and mixing whole-batch inputs with microbatch parameters is rejected.

File: torch/distributed/pipelining/schedules.py [hunk 8]
```diff
@@ -615,6 +663,63 @@ def _split_inputs(
...
+    def _get_microbatch_inputs(
+        self,
+        args: tuple[Any, ...],
+        kwargs: dict[str, Any],
+        target: Any,
+        arg_mbs: Any,
+        kwarg_mbs: Any,
+        target_mbs: Any,
+    ) -> tuple[list | None, list | None, list | None]:
+        pre_split = any(mbs is not None for mbs in (arg_mbs, kwarg_mbs, target_mbs))
```
Comment (acisseJZhong) [inline torch/distributed/pipelining/schedules.py:675]: qq: can target_mb be None while arg_mb is not None?

**Review by acisseJZhong — APPROVED** [review]

File: torch/distributed/pipelining/schedules.py (hunk 8 above)
Comment (sanketpurandare) [reply torch/distributed/pipelining/schedules.py:675]: yes, because target is only specified in the last stage
